File: typescript/test-helpers/BaseTestHelper.ts
/**
 * Base test helper utilities for refactoring exercises.
 */

/**
 * Assert that a method exists on an object.
 */
export function assertMethodExists(obj: object, methodName: string): void {
  const hasMethod = typeof (obj as any)[methodName] === 'function';
  if (!hasMethod) {
    throw new Error(`Method '${methodName}' should exist on ${obj.constructor.name}`);
  }
}

/**
 * Assert that a method is private (starts with _ or is not enumerable).
 */
export function assertMethodIsPrivate(obj: object, methodName: string): void {
  const descriptor = Object.getOwnPropertyDescriptor(obj, methodName) ||
                    Object.getOwnPropertyDescriptor(Object.getPrototypeOf(obj), methodName);
  
  if (!descriptor) {
    throw new Error(`Method '${methodName}' does not exist`);
  }

  // In TypeScript, private methods often start with _ by convention
  const isPrivateByNaming = methodName.startsWith('_');
  const isNotEnumerable = !descriptor.enumerable;
  
  if (!isPrivateByNaming && descriptor.enumerable) {
    throw new Error(`Method '${methodName}' should be private (start with _ or be non-enumerable)`);
  }
}

/**
 * Assert that an object has all expected properties.
 */
export function assertHasProperties<T extends object>(
  obj: T,
  expectedProperties: (keyof T)[],
  message?: string
): void {
  for (const prop of expectedProperties) {
    if (!(prop in obj)) {
      throw new Error(message || `Object should have property '${String(prop)}'`);
    }
  }
}

/**
 * Assert that a number is within a certain range.
 */
export function assertWithinRange(
  expected: number,
  actual: number,
  delta: number = 0.01,
  message?: string
): void {
  const diff = Math.abs(expected - actual);
  if (diff > delta) {
    throw new Error(
      message || 
      `Expected ${actual} to be within ${delta} of ${expected}, but difference was ${diff}`
    );
  }
}

/**
 * Mock function factory for testing.
 */
export function createMockFunction<T extends (...args: any[]) => any>(): T & {
  calls: Parameters<T>[];
  returns: ReturnType<T>[];
  mockReturnValue: (value: ReturnType<T>) => void;
} {
  const calls: Parameters<T>[] = [];
  const returns: ReturnType<T>[] = [];
  let returnValue: ReturnType<T>;

  const mockFn = ((...args: Parameters<T>) => {
    calls.push(args);
    const result = returnValue;
    returns.push(result);
    return result;
  }) as T & {
    calls: Parameters<T>[];
    returns: ReturnType<T>[];
    mockReturnValue: (value: ReturnType<T>) => void;
  };

  mockFn.calls = calls;
  mockFn.returns = returns;
  mockFn.mockReturnValue = (value: ReturnType<T>) => {
    returnValue = value;
  };

  return mockFn;
}

/**
 * Deep clone utility for test data.
 */
export function deepClone<T>(obj: T): T {
  if (obj === null || typeof obj !== 'object') {
    return obj;
  }

  if (obj instanceof Date) {
    return new Date(obj.getTime()) as unknown as T;
  }

  if (Array.isArray(obj)) {
    return obj.map(item => deepClone(item)) as unknown as T;
  }

  const cloned = {} as T;
  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      cloned[key] = deepClone(obj[key]);
    }
  }

  return cloned;
}

File: python/test_helpers/base_test_helper.py
"""Base test helper utilities for refactoring exercises."""

import inspect
from typing import Any, Dict, List, TypeVar, Union
import copy

T = TypeVar("T")


def assert_method_exists(obj: Any, method_name: str) -> None:
    """Assert that a method exists and is callable on an object."""
    if not hasattr(obj, method_name):
        raise AssertionError(
            f"Method '{method_name}' should exist on {obj.__class__.__name__}"
        )

    method = getattr(obj, method_name)
    if not callable(method):
        raise AssertionError(
            f"Method '{method_name}' should be callable on {obj.__class__.__name__}"
        )


def assert_method_is_private(obj: Any, method_name: str) -> None:
    """Assert that a method is private (starts with _ by Python convention)."""
    if not hasattr(obj, method_name):
        raise AssertionError(f"Method '{method_name}' does not exist")

    if not method_name.startswith("_"):
        raise AssertionError(f"Method '{method_name}' should be private (start with _)")


def assert_has_attributes(
    obj: Any, expected_attributes: List[str], message: str = ""
) -> None:
    """Assert that an object has all expected attributes."""
    for attr in expected_attributes:
        if not hasattr(obj, attr):
            error_msg = message or f"Object should have attribute '{attr}'"
            raise AssertionError(error_msg)


def assert_within_range(
    expected: float, actual: float, delta: float = 0.01, message: str = ""
) -> None:
    """Assert that a number is within a certain range."""
    diff = abs(expected - actual)
    if diff > delta:
        error_msg = (
            message
            or f"Expected {actual} to be within {delta} of {expected}, but difference was {diff}"
        )
        raise AssertionError(error_msg)


def assert_dict_contains_keys(
    expected_keys: List[str], actual_dict: Dict[str, Any], message: str = ""
) -> None:
    """Assert that a dictionary contains all expected keys."""
    for key in expected_keys:
        if key not in actual_dict:
            error_msg = message or f"Dictionary should contain key '{key}'"
            raise AssertionError(error_msg)


class MockFunction:
    """Simple mock function for testing."""

    def __init__(self, return_value: Any = None) -> None:
        self.calls: List[tuple] = []
        self.returns: List[Any] = []
        self._return_value = return_value

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        self.calls.append((args, kwargs))
        self.returns.append(self._return_value)
        return self._return_value

    def mock_return_value(self, value: Any) -> None:
        """Set the return value for future calls."""
        self._return_value = value

    @property
    def call_count(self) -> int:
        """Number of times the function was called."""
        return len(self.calls)

    def assert_called_with(self, *expected_args: Any, **expected_kwargs: Any) -> None:
        """Assert the function was called with specific arguments."""
        expected_call = (expected_args, expected_kwargs)
        if expected_call not in self.calls:
            raise AssertionError(
                f"Expected call {expected_call} not found in {self.calls}"
            )


def deep_copy(obj: T) -> T:
    """Deep copy utility for test data."""
    return copy.deepcopy(obj)


def get_method_line_count(obj: Any, method_name: str) -> int:
    """Get the number of lines in a method (useful for testing refactoring)."""
    if not hasattr(obj, method_name):
        raise ValueError(
            f"Method '{method_name}' does not exist on {obj.__class__.__name__}"
        )

    method = getattr(obj, method_name)
    try:
        source_lines = inspect.getsourcelines(method)[0]
        # Filter out empty lines and comments
        code_lines = [
            line
            for line in source_lines
            if line.strip() and not line.strip().startswith("#")
        ]
        return len(code_lines)
    except OSError:
        # Method might be built-in or defined dynamically
        return 0


def assert_method_line_count_below(obj: Any, method_name: str, max_lines: int) -> None:
    """Assert that a method has fewer than a certain number of lines."""
    line_count = get_method_line_count(obj, method_name)
    if line_count > max_lines:
        raise AssertionError(
            f"Method '{method_name}' has {line_count} lines, should be <= {max_lines}"
        )
